[#47599] Changes project archive process to only archive active subprojects

DubeySandeep opened the pull request:
## Description:

This PR fixes https://community.openproject.org/projects/openproject/work_packages/47599

Currently, archiving a project requires rights to archive the project and subprojects ([ref](https://github.com/opf/openproject/blob/ee61b6d2b3ce9d19ad0a0a08feaf5a123869ec4a/app/contracts/projects/archive_contract.rb#L65)). Also, users are not allowed to take actions on an archive/active=false project. ([ref](https://github.com/opf/openproject/blob/ee61b6d2b3ce9d19ad0a0a08feaf5a123869ec4a/app/services/authorization/user_allowed_service.rb#L82))

Through this PR  we are changing the contracts to allow archiving of a project when the users have the rights to archive the project and **active** subprojects.

## Proof of the changes:
[scrnli_14_04_2023_06-32-17.webm](https://user-images.githubusercontent.com/16653571/231914728-73f38f37-6867-420b-958f-ec70e05711b1.webm)

DubeySandeep committed:
Changes project archive process to only archive active subprojects

DubeySandeep commented on app/services/projects/archive_service.rb:
```diff
@@ -48,7 +48,8 @@ def persist(service_call)
       # We do not care for validations but want the timestamps to be updated
-      project.update_attribute(:active, false)
+      # for the changed models.
+      project.update_attribute(:active, false) if project.active?
```
Not sure whether these changes are required i.e, avoid updating the timestamp if the models are already archived. 

Note: These changes are not required to fix the issue but I've added this in case we don't want to update the timestamp of the models by archiving already archived projects.

Alternatively, we can only call `archive_project` for the subprojects which are active (by changing on line 43 of this file.)

wielinde replied:
Hey @DubeySandeep, I think it is a good idea to change line 42 to only iterate over active projects and not change this method here.

And while you are already at line 42, do you mind to change it to iterate over all active subprojects and not only direct children (maybe you also meant that?). I think that would make it consistent with the permission checks in the contract which also check for subprojects and not only children. However, if you decide to change that please also cover that new behavior in the spec archiving a project hierarchy with a depth of at least 3.

DubeySandeep replied:
> I think it is a good idea to change line 42 to only iterate over active projects and not change this method here.

Thanks for the suggestion, I've updated the changes accordingly.

>  do you mind changing it to iterate over all active subprojects and not only direct children (maybe you also meant that?)

Yes, I've updated it to cover all the subprojects. (I had found this issue "only archiving next children" while testing.)

> I think that would make it consistent with the permission checks in the contract which also check for subprojects and not only children. However, if you decide to change that please also cover that new behavior in the spec archiving a project hierarchy with a depth of at least 3.


I agree, though I'm worried there might exist some subprojects in the datastore that users have assumed to be archived but they are still active. Also, I don't think we can run a one-off job to archive all such subprojects as (assuming) a subproject can be manually unarchived by a user, right? (I don't have full context here so leaving my thoughts in case it helps decide whether we need to take any actions for such subprojects.)


Also, I've added the test for the new functionality.

DubeySandeep committed:
Fix lint issue in archive_contract.rb

wielinde reviewed (changes requested):
Hi @DubeySandeep! Thank you for the PR. Great that you have already added specs.

As this is a clear bug fix, could you please rebase your branch on `release/12.5` and then change this PRs target branch accordingly? Then the fix would get released already with the next patch level release.

Looking at that part of the code I found more problems that were not addressed in the original bug report. The original code checks the permissions for all subprojects but only archives the current project and its direct children. I find that inconsistent. It should archive all suprojects and not only direct children. If you have the time, may I ask you to also fix that and cover it with some specs?

DubeySandeep committed:
Fix archive project procces to include all subprojects

DubeySandeep committed:
Fix style & nits

DubeySandeep commented:
> As this is a clear bug fix, could you please rebase your branch on release/12.5 and then change this PRs target branch accordingly? Then the fix would get released already with the next patch level release.

Ah, I was not aware of it. Thanks for the info, I've updated the PR!

> Looking at that part of the code I found more problems that were not addressed in the original bug report. The original code checks the permissions for all subprojects but only archives the current project and its direct children. I find that inconsistent. It should archive all suprojects and not only direct children. If you have the time, may I ask you to also fix that and cover it with some specs?

@wielinde Thanks for the review, I've updated the changes accordingly and added the required test, PTAL!

DubeySandeep commented on spec/services/projects/archive_service_spec.rb:
```diff
@@ -0,0 +1,128 @@
+
+require 'spec_helper'
+
+describe Projects::ArchiveService do
```
I'm new to rspec, so I may not have followed the best practices for writing a new test file for the entire service. I'm open to suggestions to rewrite/improve it. (Like, I'm worried whether `reload` calls are too expensive.)

Test profile:
![image](https://user-images.githubusercontent.com/16653571/231913814-93fb3820-eb20-4846-83aa-be02a6e8bf85.png)

wielinde commented:
Thank you, @DubeySandeep, for your contributions. I will be offline for a couple of days and will be able to review the PR on Wednesday.

DubeySandeep committed:
Address rubospec errors

oliverguenther committed:
Don't filter for active in ruby

oliverguenther reviewed (approved):
LGTM. I've pushed a minor change to filter the active subprojects in DB instead of doing that in Ruby.
